model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProvider.java

```diff
@@ -46,18 +48,19 @@ public void close() {
     @Override
     @SuppressWarnings("unchecked")
     public <V extends AbstractEntity, M> MapStorage<V, M> getStorage(Class<M> modelType, Flag... flags) {
-        factory.validateAndUpdateSchema(session, modelType);
+        // validate and update the schema for the storage.
+        this.factory.validateAndUpdateSchema(this.session, modelType);
+        // create the JPA transaction and enlist it if needed.
+        if (session.getAttribute(this.sessionTxKey) == null) {
+            KeycloakTransaction jpaTransaction = new JpaTransactionWrapper(em.getTransaction());
+            session.getTransactionManager().enlist(jpaTransaction);
+            session.setAttribute(this.sessionTxKey, jpaTransaction);
```

Comment:
Should this be under `synchronized` guard, like e.g. here? https://github.com/keycloak/keycloak/blob/b356618cc2e00ef3ae37f515c1d4b3bdbcde3f62/model/jpa/src/main/java/org/keycloak/connections/jpa/DefaultJpaConnectionProviderFactory.java#L160-L162

Reply:
I'm not sure about this one. I think this boils down to the question: can `getStorage` be called concurrently within the same session?